### PR TITLE
feat: expose yaml and cli configuration to user test cases/suites

### PR DESCRIPTION
**Pull request by BKaDamien**
Expose all configuration parameters to test cases and suites.

Just try to make it in few lines and flexible.

In addition I don't want to create additionnal data structures so I respect the one from the entry point (see main method) and the yaml.

Finally, from my point of view, this is not a POC.

### CODE REVIEW

**Review by sebastianpfischer — APPROVED** [review]

**Review by Pog3k — APPROVED** [review]

File: docs/advanced_features/access_config.rst [hunk 1]
```diff
@@ -0,0 +1,30 @@
+Access framework's configuration
+--------------------------------
+
+All parameters given at CLI and yaml level are available for each test cases/suites.
+For a convenient usage, all configuration information are class based represented.
+This means each parameters is accessible like a "normal" instance attribute and
+the attribute's name is simply the one given in the yaml or the CLI level.
+
+Each parameters store in GlobalConfig's yaml and CLI attributes are read-only.
```
Comment (sebclrsn) [inline docs/advanced_features/access_config.rst:9]:
```suggestion
Each parameter stored in GlobalConfig's yaml and CLI attributes is read-only.
```

File: docs/advanced_features/access_config.rst [hunk 2]
```diff
@@ -0,0 +1,30 @@
+Access framework's configuration
+--------------------------------
+
+All parameters given at CLI and yaml level are available for each test cases/suites.
+For a convenient usage, all configuration information are class based represented.
+This means each parameters is accessible like a "normal" instance attribute and
```
Comment (sebclrsn) [inline docs/advanced_features/access_config.rst:6]:
```suggestion
This means each parameter is accessible like a "normal" instance attribute (dot-access) and
```

File: docs/advanced_features/access_config.rst [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+the attribute's name is simply the one given in the yaml or the CLI level.
+
+Each parameters store in GlobalConfig's yaml and CLI attributes are read-only.
+
+.. warning:: assign a new value will automatically raise an AttributeError
+
+Let says, we have the following yaml configuration file:
```
Comment (sebclrsn) [inline docs/advanced_features/access_config.rst:13]:
```suggestion
Let's admit we have the following yaml configuration file:
```

File: examples/conf_access/test_access.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        logging.info(f"variant filter: {cli_config.variant}")
+        logging.info(f"branch level: {cli_config.branch_level}")
+        logging.info(f"pattern: {cli_config.pattern}")
+
+    def test_run(self):
+        """In this case the default test_run method is overridden and
+        instead of calling test_run from BasicTest class the following
+        code is called.
+
+        Here, the test pass at the 3rd attempt out of 5. The setup and
```
Comment (sebclrsn) [inline examples/conf_access/test_access.py:66]: To be removed/adapted

File: src/pykiso/global_config.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        """
+        super().__init__(**kwargs)
+
+    def __setattr__(self, attr: str, value: Any) -> None:
+        """Protect instance attributes from writing.
+
+        :param attr: attribute's name
+        :param value: value to apply
+
+        :raises AttributeError:
```
Comment (sebclrsn) [inline src/pykiso/global_config.py:67]:
```suggestion
        :raises AttributeError: when called
```

File: src/pykiso/global_config.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+        :param config: configuration's content
+
+        :return: configuration's content as an object representation
+        """
+        str_conf = json.dumps(config)
+        return json.loads(str_conf, object_hook=lambda x: ProtectedNamespace(**x))
+
+    @staticmethod
+    def grab_yaml_config(func) -> Callable:
```
Comment (sebclrsn) [inline src/pykiso/global_config.py:103]:
```suggestion
    def grab_yaml_config(func: Callable) -> Callable:
```

File: src/pykiso/global_config.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+        @functools.wraps(func)
+        def grab_inner(*args, **kwargs) -> dict:
+            """Grab the return values from yaml configuration parser
+            method and store it in the GlobalConfig instance.
+
+            :param args: positonal arguments
+            :param kwargs: named arguments
+
+            :return: parsed yam configuration information
```
Comment (sebclrsn) [inline src/pykiso/global_config.py:119]:
```suggestion
            :return: parsed yaml configuration information
```

File: src/pykiso/global_config.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+        """Initialize cli and yaml attributes by taking the content of
+        Grabber collector object.
+        """
+        self.cli = None
+        self.yaml = None
+
+
+class Grabber:
+    """Responsible to collect configuration information at different
+    level of the framework."""
```
Comment (sebclrsn) [inline src/pykiso/global_config.py:88]:
```suggestion
    levels of the framework."""
```

File: src/pykiso/global_config.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+        :param value: value to apply
+
+        :raises AttributeError:
+        """
+        raise AttributeError(f"Attribute {attr} is not writable")
+
+
+class GlobalConfig(metaclass=Singleton):
+    """Container object used by the user to have access to all
+    configuration information coming from different level of the
```
Comment (sebclrsn) [inline src/pykiso/global_config.py:74]:
```suggestion
    configuration information coming from the different levels of the
```

**Review by sebclrsn — APPROVED** [review]

File: examples/conf_access/test_access.py [hunk 10]
```diff
@@ -0,0 +1,97 @@
+##########################################################################
+# Copyright (c) 2010-2021 Robert Bosch GmbH
```
Comment (kakooloukia) [inline examples/conf_access/test_access.py:2]:
```suggestion
# Copyright (c) 2010-2022 Robert Bosch GmbH
```

**Review by kakooloukia — APPROVED** [review]

**Pull request merged**